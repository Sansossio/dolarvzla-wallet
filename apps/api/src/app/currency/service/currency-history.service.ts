import { Injectable, NotFoundException } from '@nestjs/common'
import { InjectRepository } from '@nestjs/typeorm'
import { Repository } from 'typeorm'
import { CurrencyHistoryEntity } from '@dolarvzla-wallet/models'
import { GetCurrencyPriceRequestDto } from '../dto/request/get-currency-price.dto.request'
import { GetCurrencyPriceDtoResponse } from '../dto/response/get-currency-price.dto.response'

@Injectable()
export class CurrencyHistoryService {
  constructor (
    @InjectRepository(CurrencyHistoryEntity)
    private readonly repository: Repository<CurrencyHistoryEntity>
  ) {}

  async getCurrencyPrice (query: GetCurrencyPriceRequestDto): Promise<GetCurrencyPriceDtoResponse> {
    const data = await this.repository.findOne({
      where: {
        currency: query.currency
      },
      order: {
        createAt: 'DESC'
      },
      select: [
        'price',
        'createAt'
      ]
    })

    if (!data) {
      throw new NotFoundException()
    }

    return {
      currency: query.currency,
      price: data.price,
      lastUpdate: data.createAt
    }
  }
}
