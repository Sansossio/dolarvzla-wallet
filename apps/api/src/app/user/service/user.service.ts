import { ConflictException, Injectable } from '@nestjs/common'
import { InjectRepository } from '@nestjs/typeorm'
import { Repository } from 'typeorm'
import { UserEntity } from '@dolarvzla-wallet/models'
import { RegisterUserDtoRequest } from '../dto/request/register-user.dto.request'

@Injectable()
export class UserService {
  constructor (
    @InjectRepository(UserEntity)
    private readonly repository: Repository<UserEntity>
  ) {}

  async register (body: RegisterUserDtoRequest) {
    const exists = !!await this.repository.findOne({
      where: {
        email: body.email
      },
      select: [
        'id'
      ]
    })
    if (exists) {
      throw new ConflictException('User already exists')
    }
    const newUser = this.repository.create({
      email: body.email,
      password: body.password
    })

    await this.repository.save(newUser)
  }
}
