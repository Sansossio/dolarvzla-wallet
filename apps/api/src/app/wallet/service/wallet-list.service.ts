import { Injectable } from '@nestjs/common'
import { Repository } from 'typeorm'
import { UserWalletEntity } from '@dolarvzla-wallet/models'
import { InjectRepository } from '@nestjs/typeorm'
import { WalletListResponseDto } from '../dto/response/wallet-list.response.dto'
import { WalletOrchestrator } from '@dolarvzla-wallet/crypto-wallet'

@Injectable()
export class WalletListService {
  constructor (
    @InjectRepository(UserWalletEntity)
    private readonly repository: Repository<UserWalletEntity>,

    private readonly walletOrquestador: WalletOrchestrator
  ) {}

  async list (userId: number): Promise<WalletListResponseDto[]> {
    const listOfWallets = await this.repository.find({
      where: {
        user: { id: userId }
      },
      select: [
        'type',
        'walletId',
        'address'
      ]
    })

    return Promise.all(
      listOfWallets.map(async (wallet): Promise<WalletListResponseDto> => {
        const { balance } = await this.walletOrquestador
          .getHandler(wallet.type)
          .getBalance(wallet.walletId)

        return WalletListResponseDto.fromEntity(wallet, balance)
      })
    )
  }
}
