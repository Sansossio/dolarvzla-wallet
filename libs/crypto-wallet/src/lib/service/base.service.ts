import { HttpService, Injectable } from '@nestjs/common'
import { ConfigService } from '@nestjs/config'
import { NewWalletDto } from './dto/new-wallet.dto'
import { WalletBalanceDto } from './dto/wallet-balance.dto'

@Injectable()
export abstract class BaseService {
  constructor (
    protected readonly httpService: HttpService,
    protected readonly config: ConfigService
  ) {}

  abstract createRandomWallet (): Promise<NewWalletDto>

  abstract listWallets (): Promise<string[]>

  abstract generateNewAddress (walletName: string): Promise<string>

  abstract getBalance (walletname: string): Promise<WalletBalanceDto>
}
