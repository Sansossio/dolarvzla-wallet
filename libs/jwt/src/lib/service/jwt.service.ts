import { Injectable } from '@nestjs/common'
import { ConfigService } from '@nestjs/config'
import * as jwt from 'jsonwebtoken'
import { ObjectLiteral } from 'typeorm'
import { clearToken } from './clear-token/clear-token.utils'

@Injectable()
export class JwtService {
  private readonly jwt = jwt
  private readonly secret = this.config.get<string>('jwt-crypto.jwt.secret')
  private readonly options: jwt.SignOptions = {
    expiresIn: this.config.get('jwt-crypto.jwt.expiration')
  }

  constructor (
    protected readonly config: ConfigService
  ) {}

  sign (payload: ObjectLiteral): string {
    return this.jwt.sign(payload, this.secret, this.options)
  }

  decode<T extends ObjectLiteral> (token: string): T {
    return this.jwt.verify(clearToken(token), this.secret) as T
  }

  async verify<T extends ObjectLiteral | string> (token: string, secret: jwt.Secret, options: jwt.VerifyOptions) {
    const newToken = clearToken(token)
    return new Promise<T>((resolve, reject) => {
      this.jwt.verify(newToken, secret, options, (err, data) => {
        if (err) reject(err)
        else resolve(data as T)
      })
    })
  }
}
