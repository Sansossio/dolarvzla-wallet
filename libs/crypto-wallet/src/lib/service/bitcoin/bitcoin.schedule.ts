import { Injectable } from '@nestjs/common'
import { BitcoinService } from './bitcoin.service'
import * as ms from 'ms'
import { Interval } from '@nestjs/schedule'
import { ConfigService } from '@nestjs/config'
import { AvailableMethodsRpc } from './dto/available-methods.enum'

const INTERVAL_GENERATE_BLOCKS = ms('30s')

@Injectable()
export class BitcoinSchedule {
  constructor (
    private readonly service: BitcoinService,
    private readonly config: ConfigService
  ) {}

  @Interval(INTERVAL_GENERATE_BLOCKS)
  async cronAutoGenerateBlocks () {
    const config = this.config.get('bitcoin.autoGenerateBlocks')
    if (!config.enable) {
      return
    }
    const listWallets = await this.service.listWallets()
    await Promise.all(
      listWallets.map(async (wallet) => {
        const address = await this.service.generateNewAddress(wallet)
        await (this.service as any).rpcRequest(AvailableMethodsRpc.GENERATETOADDRESS, [config.amount, address])
      })
    )
  }
}
