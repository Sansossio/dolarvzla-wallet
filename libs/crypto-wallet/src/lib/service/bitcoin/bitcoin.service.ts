import { Injectable } from '@nestjs/common'
import { BaseService } from '../base.service'
import { v4 as uuidv4 } from 'uuid'
import { NewWalletDto } from '../dto/new-wallet.dto'
import { AvailableMethodsRpc } from './dto/available-methods.enum'
import { RpcResponseDto } from './dto/rpc-response.dto'
import { WalletBalanceDto } from '../dto/wallet-balance.dto'
import { Currency } from '@dolarvzla-wallet/models'

@Injectable()
export class BitcoinService extends BaseService {
  type = Currency.BITCOIN

  private readonly rpcConfig = {
    domain: this.config.get('bitcoin.rpc.domain'),
    port: this.config.get('bitcoin.rpc.port'),
    username: this.config.get('bitcoin.rpc.username'),
    password: this.config.get('bitcoin.rpc.password')
  }

  private async rpcRequest<T> (method: AvailableMethodsRpc, params: any[] = [], path = ''): Promise<T> {
    const url = `http://${this.rpcConfig.username}:${this.rpcConfig.password}@${this.rpcConfig.domain}:${this.rpcConfig.port}/${path}`
    const {
      data: response
    } = await this.httpService.post<RpcResponseDto>(
      url,
      {
        jsonrpc: '2.0',
        id: uuidv4(),
        method,
        params
      }
    ).toPromise()
    return response.result
  }

  private getWalletPath (walletName: string) {
    return `wallet/${walletName}`
  }

  async listWallets (): Promise<string[]> {
    return (await this.rpcRequest<string[]>(AvailableMethodsRpc.LISTWALLETS)).filter(val => !!val)
  }

  async generateNewAddress (walletName: string) {
    return this.rpcRequest<string>(AvailableMethodsRpc.GETNEWADDRESS, [], this.getWalletPath(walletName))
  }

  async createRandomWallet (): Promise<NewWalletDto> {
    const walletName = uuidv4()
    const { name } = await this.rpcRequest<{ name: string }>(AvailableMethodsRpc.CREATEWALLET, [walletName])

    return {
      name,
      address: await this.generateNewAddress(walletName)
    }
  }

  async getBalance (walletName: string): Promise<WalletBalanceDto> {
    const balance = await this.rpcRequest<number>(AvailableMethodsRpc.GETBALANCE, [], this.getWalletPath(walletName))

    return {
      balance
    }
  }
}
