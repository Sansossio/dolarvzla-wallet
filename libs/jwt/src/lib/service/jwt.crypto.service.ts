import { Injectable } from '@nestjs/common'
import * as jwt from 'jsonwebtoken'
import { ObjectLiteral } from 'typeorm'
import { AesCrypto } from '../../../../crypto/src'
import { clearToken } from './clear-token/clear-token.utils'
import { JwtService } from './jwt.service'

@Injectable()
export class JwtCryptoService extends JwtService {
  private readonly aesCryptoSecret = this.config.get<string>('jwt-crypto.crypto.secret')

  sign (payload: ObjectLiteral): string {
    const token = super.sign(payload)
    return AesCrypto.encrypt(token, this.aesCryptoSecret)
  }

  decode<T extends ObjectLiteral> (token: string): T {
    token = AesCrypto.decrypt(clearToken(token), this.aesCryptoSecret)
    return super.decode(token)
  }

  async verify<T extends ObjectLiteral | string> (token: string, secret: jwt.Secret, options: jwt.VerifyOptions) {
    return super.verify<T>(
      AesCrypto.decrypt(clearToken(token), this.aesCryptoSecret),
      secret,
      options
    )
  }
}
