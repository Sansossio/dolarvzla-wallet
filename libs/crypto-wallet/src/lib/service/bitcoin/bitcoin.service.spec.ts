import { BitcoinService } from './bitcoin.service'
import { AvailableMethodsRpc } from './dto/available-methods.enum'

describe('BitcoinService', () => {
  const mockPost = jest.fn()
  const httpService = {
    post: () => ({ toPromise: mockPost })
  }
  const configService = {
    get: jest.fn()
  }

  const service = new BitcoinService(
    httpService as any,
    configService as any
  )

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should be defined', () => {
    expect(service).toBeDefined()
  })

  describe('createRandomWallet', () => {
    it('should match snapshot', async () => {
      mockPost
        .mockImplementationOnce(async () => ({ data: { result: { name: 'wallletName' } } }))
        .mockImplementationOnce(async () => ({ data: { result: 'walletAddress' } }))

      await expect(service.createRandomWallet()).resolves.toMatchSnapshot()
    })
  })

  describe('generateNewAddress', () => {
    it('should match snapshot', async () => {
      const walletName = 'MY_WALLET'
      mockPost.mockImplementation(async () => ({ data: { result: walletName } }))

      await expect(service.generateNewAddress(walletName)).resolves.toMatchSnapshot()
    })

    it('should called with empty params and wallet name', async () => {
      const walletName = 'MY_WALLET'
      const rpcRequest = jest.spyOn(service, 'rpcRequest' as any)

      await service.generateNewAddress(walletName)

      expect(rpcRequest).toBeCalledWith(AvailableMethodsRpc.GETNEWADDRESS, [], `wallet/${walletName}`)
    })
  })
})
