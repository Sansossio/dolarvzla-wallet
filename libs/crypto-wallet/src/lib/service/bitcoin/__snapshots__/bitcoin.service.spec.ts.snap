// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BitcoinService createRandomWallet should match snapshot 1`] = `
Object {
  "address": "walletAddress",
  "name": "wallletName",
}
`;

exports[`BitcoinService generateNewAddress should match snapshot 1`] = `"MY_WALLET"`;
